django_emarsys/event.py

```diff
@@ -0,0 +1,237 @@
+                num_updated_ids += 1
+
+            local_event.emarsys_id = emarsys_event_id
+            del emarsys_events[local_event.name]
+            local_event.save()
+
+    for emarsys_event_name, emarsys_event_id in emarsys_events.items():
+        if emarsys_event_name not in known_event_names:
+            continue
+
+        Event.objects.create(emarsys_id=emarsys_event_id,
+                             name=emarsys_event_name)
+        num_new_events += 1
+
+    all_synced_event_names = set(Event.objects.all()
+                                 .values_list('name', flat=True))
+    unsynced_event_names = list(known_event_names - all_synced_event_names)
+
+    if unsynced_event_names:
+        log.warning("these event names in settings.EMARSYS_EVENTS are not "
+                    "known by Emarsys: {}"
+                    .format(', '.join(
+                        '"{}"'.format(x.replace('"', '\\"'))
+                            for x in unsynced_event_names)))
+
+    return num_new_events, num_updated_ids, num_deleted_ids, \
+        unsynced_event_names
+
+
+def get_placeholder_data(event_name, **kwargs):
+    context_provider = get_context_provider(event_name)
+
+    return context_provider(**kwargs)
+
+
+def trigger_event(event_name, recipient_email, data=None,
+                  create_user_if_needed=True,
+                  contact_data_provider=None,
+                  manual=False):
+    if manual:
+        source = EventInstance.SOURCE_MANUAL
+    else:
+        source = EventInstance.SOURCE_AUTOMATIC
+
+    emarsys_event_id = get_event_id(event_name)
+
+    event = _create_event_instance(
+        event_name=event_name,
+        recipient_email=recipient_email,
+        emarsys_event_id=emarsys_event_id,
+        source=source,
+        data=data)
+
+    if (event.state == EventInstance.STATE_ERROR
+            and event.result_code == '2008'
+            and create_user_if_needed):
+
+        if contact_data_provider is not None:
+            contact_data = contact_data_provider()
+        else:
+            contact_data = {'E-Mail': recipient_email}
+
+        api.create_contact(contact_data)
+        event = _create_event_instance(
+            event_name=event_name,
+            recipient_email=recipient_email,
+            emarsys_event_id=emarsys_event_id,
+            source=source,
+            data=data)
+
+    return event
+
+
+def _create_event_instance(event_name, recipient_email, emarsys_event_id,
+                           source, data):
+    """
+    A `EventInstance` object is instantiated. All data validation is
+    done here, passing along correct user `data`.
+
+    Upon any error the resulting event instance's will store the error message
+    and get the appropriate state.
+
+    :returns: the new event object
+
+    """
+    event = EventInstance.objects.create(
+        event_name=event_name,
+        recipient_email=recipient_email,
+        source=source,
+        emarsys_id=emarsys_event_id)
```

Comment:
Hmm, this database write probably never sees the light of a database commit: In the success case, `handle_success` is called, which writes again, making the first write unnecessary. In expected error cases, `handle_error` is called, which writes again. In unexpected error cases, the transaction is very likely to be rollbacked.

So work with an unsaved instance instead?